Segfault on SFrame when file not exists

This sefault has nothing to do with S3. `is_directory` throws an exception and, that's it.

segfaults on internal build eggs through `pip install`,
```
Current thread 0x00000001141c15c0 (most recent call first):
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/turicreate/data_structures/sframe.py", line 2546 in column_types
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/turicreate/data_structures/sframe.py", line 2078 in __get_column_description__
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/turicreate/data_structures/sframe.py", line 2062 in __repr__
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_io/saferepr.py", line 56 in repr_instance
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/reprlib.py", line 65 in repr1
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/reprlib.py", line 55 in repr
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_io/saferepr.py", line 47 in repr
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_io/saferepr.py", line 82 in saferepr
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_code/code.py", line 693 in repr_args
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_code/code.py", line 784 in repr_traceback_entry
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_code/code.py", line 823 in repr_traceback
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_code/code.py", line 879 in repr_excinfo
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/_code/code.py", line 634 in getrepr
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/nodes.py", line 361 in _repr_failure_py
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/python.py", line 1513 in repr_failure
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/reports.py", line 294 in from_item_and_call
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/runner.py", line 260 in pytest_runtest_makereport
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/runner.py", line 188 in call_and_report
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/runner.py", line 100 in runtestprotocol
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/runner.py", line 85 in pytest_runtest_protocol
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/main.py", line 272 in pytest_runtestloop
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/main.py", line 247 in _main
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/main.py", line 191 in wrap_session
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/_pytest/config/__init__.py", line 125 in main
  File "/Users/znation/builds/2-cVbmd6/0/turi/turicreate/deps/env/lib/python3.6/site-packages/pytest/__main__.py", line 7 in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85 in _run_code
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193 in _run_module_as_main
```

But this won't happen in a local debug build environment. Here's how you repro

1. create a virtual env
2. pip install the wheel
3. copy the turicreate directory to source development repo /debug/src/python
4. run pytest -s turicreate/test/test_sframe_s3.py -k download

you can find no segfault is there. Strange issue.

in my local build, it throws an error instead of a segfault.

```
================================================================================================================== FAILURES ==================================================================================================================
___________________________________________________________________________________________ TestSFrameS3.test_s3_sframe_download[small_sframe_dc] ____________________________________________________________________________________________

self = <turicreate.test.test_sframe_s3.TestSFrameS3 object at 0x16eaff860>, folder = 'small_sframe_dc'

    @pytest.mark.parametrize("folder", remote_sframe_folders)
    def test_s3_sframe_download(self, folder):
        sf_from_disk = self.downloaded_files[folder + "_sframe"]
        obj_key = os.path.join(self.s3_sframe_prefix, folder)
        s3_url = os.path.join("s3://", self.bucket, obj_key)
>       sf_from_s3 = SFrame(s3_url)

test/test_sframe_s3.py:105:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
data_structures/sframe.py:835: in __init__
    self.__proxy__.load_from_sframe_index(url)
cy_sframe.pyx:71: in turicreate._cython.cy_sframe.UnitySFrameProxy.load_from_sframe_index
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   OSError: s3://tc_qa/integration/manual/sframes/small_sframe_dc not found. ErrMsg: s3://tc_qa/integration/manual/sframes/small_sframe_dc has no objects or diretoires. Consider create the prefix and try again. operator() from /Users/zac
h/builds/1fUgszd-/0/turi/turicreate/src/core/storage/sframe_interface/unity_sframe.cpp at 122
```

--- a/src/core/storage/sframe_interface/unity_sframe.cpp
+++ b/src/core/storage/sframe_interface/unity_sframe.cpp
@@ -1856,6 +1856,8 @@ void unity_sframe::delete_on_close() {
 }
 
 std::shared_ptr<planner_node> unity_sframe::get_planner_node() {
+  ASSERT_MSG(m_planner_node != nullptr,
+             "Unintialized SFrame planner node cannot be used for read");
   return m_planner_node;
 }
 

--- a/src/python/turicreate/test/test_io_s3.py
+++ b/src/python/turicreate/test/test_io_s3.py
@@ -129,6 +129,20 @@ class TestSFrameS3(object):
         # we can trust the upload becuase if the upload fails,
         # s3 will respond with 5xx
 
+    @pytest.mark.parametrize(
+        "url",
+        ["s3://gui/dummy", "./willy-nily/blah", "https://foo.com", "http://hao.com"],
+    )
+    def test_s3_sframe_load_from_wrong_path(self, url):
+        # s3 only writes when it receives all parts
+        # it's sort of atmoic write on file level.
+        if six.PY2:
+            with pytest.raises(IOError):
+                SFrame(url)
+        else:
+            with pytest.raises(OSError):
+                SFrame(url)
+
     def test_s3_sframe_upload_throw(self):
         # s3 only writes when it receives all parts
         # it's sort of atmoic write on file level.

